libraries/lib-telemetry/TelemetryManager.cpp

```diff
@@ -0,0 +1,430 @@
+            [index = NextSubmissionIndex, this](details::SubmissionResult result) {
+
+            std::lock_guard<std::mutex> lock (TelemetryMutex);
+
+            if (result == details::SubmissionResult::Success || 
+                result == details::SubmissionResult::Failed)
+            {
+                QueuedEvents.erase (std::remove_if (QueuedEvents.begin (), QueuedEvents.end (), 
+                    [index](const details::Event& evt) {
+                    return evt.SumbmissionIndex == index;
+                }), QueuedEvents.end ());
+
+                persistTelemetryData ();
+            }
+            else
+            {
+                for (details::Event& evt : QueuedEvents)
+                {
+                    if (evt.SumbmissionIndex == index)
+                        evt.Submitted = false;
+                }
+            }
+        });
+
+        for (size_t idx = 0;
+            firstUnqueuedEventIt != QueuedEvents.end () && idx < consumedEvents;
+            ++firstUnqueuedEventIt, ++idx)
+        {
+            if (firstUnqueuedEventIt->Submitted)
+                continue;
+
+            firstUnqueuedEventIt->SumbmissionIndex = NextSubmissionIndex;
+            firstUnqueuedEventIt->Submitted = true;
+        }
+
+        NextSubmissionIndex++;
+        enqueueTimer (firstUnqueuedEventIt != QueuedEvents.end ());
+    }
+
+    void generateClientID ()
+    {
+        ClientID = Uuid::Generate ().toString ();
+
+        std::lock_guard<std::mutex> lock (TelemetryMutex);
+
+        persistTelemetryData ();
+    }
+
+    bool readTelemetryData ()
+    {
+        wxString fileName = wxFileName (
+            wxString::FromUTF8 (TempDir),
+            telemetryFileName
+        ).GetFullPath ();
+
+        if (!wxFileExists (fileName))
+            return false;
+
+        wxFile inputFile (fileName, wxFile::read);
+
+        if (!inputFile.IsOpened ())
+            return false;
+
+        wxString content;
+
+        if (!inputFile.ReadAll (&content) || content.empty ())
+            return false;
+
+        wxStringTokenizer tokenizer (content, "\r\n");
+
+        ClientID = tokenizer.GetNextToken ();
+
+        std::lock_guard<std::mutex> lock (TelemetryMutex);
+
+        while (tokenizer.HasMoreTokens ())
+        {
+            wxString evtString = tokenizer.GetNextToken ();
+
+            if (evtString.empty ())
+                continue;
+
+            details::Event evt;
+
+            if (evt.deserialize (evtString))
+                QueuedEvents.push_back (evt);
+        }
+
+        return true;
+    }
+
+    bool persistTelemetryData ()
+    {
+        wxString fileName = wxFileName (
+            wxString::FromUTF8 (TempDir),
+            telemetryFileName
+        ).GetFullPath ();
+
+        wxFile inputFile (fileName, wxFile::write);
+
+        if (!inputFile.IsOpened ())
+            return false;
+
+        if (!inputFile.Write (wxString::FromUTF8 (ClientID + "\n")))
+            return false;
```

Comment:
I'm afraid you cannot claim anonymous tracking and do that in the same pull request.

Reply:
Can someone translate what this does to English?

Reply:
This is saving some sort of unique identifier of the user.

Reply:
Even if this was removed, Google could easily de-pseudonymize your IP. So anyone claiming "it's anonymous" just shows that they don't know what they are talking about OR are intentionally deceiving you.

Reply:
Google and Yandex also track IP addresses on their side.